### DISCUSSION TITLE
Getting States from a Theorem without Tracing

### DISCUSSION THREAD

**Discussion by princesspeace** [lean-dojo/LeanDojo#141] [Q&A]
I'm trying to extract all the intermediate states from a theorem.

Is there anyway to do this without tracing the entire repo it is in and creating an AST? 

Also, is it possible to extract the intermediate states of a theorem where has_tactic_proof() is false? Many theorems are a combination of terms and tactics.

**Comment by yangky11** [comment]
Hi,

Currently, LeanDojo does not support extracting information about a single theorem without tracing the entire repo.

Also, we only extract states before/after tactics (`has_tactic_proof()` may or may not be true). If you'd like to extract all states, I'd suggest writing a Lean metaprogram similar to what we have done here: 
https://github.com/lean-dojo/LeanDojo/blob/d399420472de62e8e544d826101585b5fa1f1ee7/src/lean_dojo/data_extraction/ExtractData.lean#L293

**Comment by princesspeace** [comment]
Hi, thanks for the response.

Would this be done just as simple as removing the ```match i.stx.getKind with``` ? I'm also curious as to why you don't extract all states - shouldn't these states contain some useful information for training ?

**Reply by yangky11** [reply]
Kind of, though I'm not 100% sure since I didn't try. In LeanDojo, we focus on tactics as the interaction mode between human/LLMs and Lean. Therefore, we only extract states before/after tactics.

**Reply by realharryhero** [reply]
Any progress on this @princesspeace ? Ideally, it'd be nice to enter the dojo just given the theorem statement in LEAN.

**Reply by darabos** [reply]
Check out [Lean REPL](https://github.com/leanprover-community/repl). It's much lower level than LeanDojo, and lacks most (all? 😅 ) of its conveniences. But it lets you just pass in a theorem as a string with a `sorry` proof and see and interact with the proof state at the `sorry`.

**Reply by realharryhero** [reply]
Sorry for the (very) late reply; that's exactly what I need. Thank you!
